File: core/src/com/colorflow/music/CircularBuffer.java
package com.colorflow.music;

public class CircularBuffer {

    private double[][] data;
    private int rows, columns;
    private int circularIndex;

    public CircularBuffer(int rows, int columns) {
        this.data = new double[rows][columns];
        this.rows = rows;
        this.columns = columns;
        this.circularIndex = 0;
    }

    public int getNumRows() {
        return rows;
    }

    public int getNumColumns() {
        return columns;
    }

    public int getRowIndex() {
        return circularIndex;
    }

    public void incRowIndex() {
        circularIndex ++;
        circularIndex %= rows;
    }

    public double[] getRow(int row) {
        return data[row].clone();
    }

    public double[] getColumn(int column) {
        double[] col = new double[rows];
        for(int i = circularIndex; i < circularIndex + rows; i++) {
            col[i - circularIndex] = data[i % rows][column];
        }
        return col;
    }

    public double[][] getMatrix() {
        return data;
    }
}

File: core/src/com/colorflow/music/CircularBuffer.kt
package com.colorflow.music

class CircularBuffer(val numRows: Int, val numColumns: Int) {

    val matrix: Array<DoubleArray> = Array(numRows) { DoubleArray(numColumns) }
    var rowIndex: Int = 0
        private set

    fun incRowIndex() {
        rowIndex++
        rowIndex %= numRows
    }

    fun getRow(row: Int): DoubleArray {
        return matrix[row].clone()
    }

    fun getColumn(column: Int): DoubleArray {
        val col = DoubleArray(numRows)
        for (i in rowIndex until rowIndex + numRows) {
            col[i - rowIndex] = matrix[i % numRows][column]
        }
        return col
    }
}
